libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -994,53 +1009,52 @@ void _IotMqtt_DecrementConnectionReferences( _mqttConnection_t * pMqttConnection
 IotMqttError_t IotMqtt_Init( void )
 {
     IotMqttError_t status = IOT_MQTT_SUCCESS;
+    uint32_t allowInitialization = Atomic_CompareAndSwap_u32( &_initCalled,
+                                                              MQTT_LIBRARY_INITIALIZED,
+                                                              MQTT_LIBRARY_UNINITIALIZED );
 
-    if( _initCalled == 0 )
+    if( allowInitialization == 1 )
     {
         /* Call any additional serializer initialization function if serializer
          * overrides are enabled. */
         #if IOT_MQTT_ENABLE_SERIALIZER_OVERRIDES == 1
             #ifdef _IotMqtt_InitSerializeAdditional
                 if( _IotMqtt_InitSerializeAdditional() == false )
                 {
+                    /* Log initialization status. */
+                    IotLogError( "Failed to initialize MQTT library serializer. " );
+
                     status = IOT_MQTT_INIT_FAILED;
+
+                    IOT_GOTO_CLEANUP();
```

Comment:
Combining these to `IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_INIT_FAILED );` would be cleaner.